### PR TITLE
BPE Fixes and tests

### PR DESCRIPTION
**Pull request by voropz**
Signed-off-by: Pavel Voropaev <8759144+voropz@users.noreply.github.com>

### CODE REVIEW

**Comment by voropz** [comment]
@nismohl2 @masepi

File: NeoML/include/NeoML/TraditionalML/SubwordEncoder.h [hunk 1]
```diff
@@ -52,6 +57,7 @@ class NEOML_API ISubwordEncoderWithCache : public ISubwordEncoder {
 	// To completely switch the cache off set cachePeriod equal to -1.
 	// Value 0 is treated as invalid.
 	void SetCachePeriod( int cachePeriod ) const { cache.SetCachePeriod( cachePeriod ); }
+	void ClearCache() const { cache.Clear(); }
```
Comment (nismohl2) [inline NeoML/include/NeoML/TraditionalML/SubwordEncoder.h:65]: A comment is necessary for public methods.
The same for CCache::Clear.

File: NeoML/src/TraditionalML/BytePairEncoder.cpp [hunk 2]
```diff
@@ -175,6 +164,37 @@ void CBytePairEncoder::Serialize( CArchive& archive )
...
+		NeoAssert( !tokenToId.Has( newToken ) );
+		tokenToId.Add( newToken, tokens.Size() );
+		tokens.Add( newToken );
+	}
+}
+
+void CBytePairEncoder::GetDictionary( CWordDictionary& output ) const
+{
+	output.Empty();
+	output.AddWord( UnkToken, 0 );
```
Comment (nismohl2) [inline NeoML/src/TraditionalML/BytePairEncoder.cpp]: Maybe 0 can be replaced by tokens.Size() + 1? And then we can count towards 0.

So we can save the property of non-negative frequencies without `NeoAssert` statements deletion (See below).

File: NeoML/src/TraditionalML/WordDictionary.cpp [hunk 3]
```diff
@@ -141,7 +139,6 @@ long long CWordDictionary::GetWordUseCount( const CString& word ) const
 double CWordDictionary::GetWordFrequency( int id ) const
 {
 	checkId( id );
-	NeoAssert( totalWordsUseCount > 0 );
```
Comment (nismohl2) [inline NeoML/src/TraditionalML/WordDictionary.cpp]: potential division by zero

File: NeoML/include/NeoML/TraditionalML/SubwordEncoder.h [hunk 4]
```diff
@@ -96,11 +103,18 @@ class NEOML_API ISubwordEncoderWithCache : public ISubwordEncoder {
...
+DECLARE_NEOML_MODEL_NAME( BytePairEncoderModelName, "NeoMLBytePairEncoderModel" )
+
 class NEOML_API IBytePairEncoder : public ISubwordEncoderWithCache {
 public:
 	// Returns encoder flags.
 	virtual bool UseEndOfWordToken() const = 0;
 	virtual bool UseStartOfWordToken() const = 0;
+
+	// Initializes the encoder. Can be safely used only once.
+	virtual void LoadDictionary( const CWordDictionary& tokens, bool useEndOfWordToken, bool useStartOfWordToken ) = 0;
```
Comment (nismohl2) [inline NeoML/include/NeoML/TraditionalML/SubwordEncoder.h]: With public `LoadDictionary` we introduce possibility when the content of `tokens` is inconsistent with values of `useEndOfWordToken` and `useStartOfWordToken`.
The initial idea was to encapsulate build process into Trainer class, so we can guarantee that resulting Encoder is consistent and valid.

File: NeoML/src/TraditionalML/WordDictionary.cpp [hunk 5]
```diff
@@ -96,10 +92,12 @@ void CWordDictionary::Finalize( long long minCount )
 {
 	words.QuickSort<Descending<CWordWithCount>>();
 
-	const CWordWithCount countSeparator{ "", minCount - 1 };
-	const int insertionPoint = words.FindInsertionPoint<Descending<CWordWithCount>>( countSeparator );
+	if( minCount > INT64_MIN ) {
```
Comment (nismohl2) [inline NeoML/src/TraditionalML/WordDictionary.cpp:98]: Unnecessary in case of non-negative frequencies.
But check for `minCount` positivity will be useful.

**Review by nismohl2 — CHANGES_REQUESTED** [review]
Need additional discussion on topics:
1. Negative frequencies in `CWordDictionary`
2. Public access to `LoadDictionary` can lead to invalid Encoder.

File: NeoML/include/NeoML/TraditionalML/SubwordEncoder.h (hunk 4 above)
Comment (NValerij) [reply NeoML/include/NeoML/TraditionalML/SubwordEncoder.h]: `LoadDictionary`  can be useful for experiments. For example, we want to try external (sentence-piece trained on another corpus) vocabulary. Looks like we can easily check `tokens` for our special subtokens and raise exception if any exists.

File: NeoML/Python/src/PyBytePairEncoder.cpp [hunk 6]
```diff
@@ -63,8 +74,50 @@ void CPyBytePairEncoder::Train( py::dict dictionary, int tokensCount,
 	}
 }
 
+py::dict CPyBytePairEncoder::GetDictionary() const
+{
+	NeoAssert( encoder != nullptr );
+
+	py::dict result;
+	CWordDictionary dictionary;
+	encoder->GetDictionary( dictionary );
```
Comment (FedyuninV) [inline NeoML/Python/src/PyBytePairEncoder.cpp]:
```
{
    py::gil_scoped_release release;
    encoder->GetDictionary( dictionary );
}
```

File: NeoML/docs/ru/API/Algorithms/README.md [hunk 7]
```diff
@@ -455,3 +461,10 @@ public:
 2. Создать экземпляр класса `CBytePairEncoderTrainer` с желаемыми параметрами `CParams` и ранее созданным словарем.
 3. Вызвать метод `Train` у экземпляра `CBytePairEncoderTrainer`.
     * Воспользоваться методом `TrainSteps`, если есть необходимость создать частично обученный кодировщик. Получить его можно с помощью вызова метода `GetEncoder`.
+
+В целях отладки `IBytePairEncoder` предоставляет прямые методы загрузки и выгрузки словаря (`LoadDictionary` и `GetDictionary`). Словарь, созданный не алгоритмами NeoML, должен соотве:
```
Comment (FedyuninV) [inline NeoML/docs/ru/API/Algorithms/README.md]: `должен соотве:`

**Review by FedyuninV — COMMENTED** [review]
1. Mention BPE [here](https://github.com/neoml-lib/neoml/blob/master/NeoML/docs/en/Python/submodules/misc.rst) with corresponding `..autoclass`
2. Add some simplest Python tests [here](https://github.com/neoml-lib/neoml/blob/master/NeoML/Python/tests.py) . The main purpose of these tests is to check that Python wrapper is working as intended. Tests for the algorithm, corner cases, etc. aren't needed there.

**Review by FedyuninV — APPROVED** [review]
